src/bench/addrman.cpp

```diff
@@ -95,6 +110,43 @@ static void AddrManSelect(benchmark::Bench& bench)
     });
 }
 
+// The worst case performance of the Select() function is when there is only
+// one address on the table, because it linearly searches every position of
+// several buckets before identifying the correct bucket
+static void AddrManSelectFromAlmostEmpty(benchmark::Bench& bench)
+{
+    AddrMan addrman{EMPTY_NETGROUPMAN, /*deterministic=*/false, ADDRMAN_CONSISTENCY_CHECK_RATIO};
+
+    // Add one address to the new table
+    CService addr = ResolveService("250.3.1.1", 8333);
+    addrman.Add({CAddress(addr, NODE_NONE)}, ResolveService("250.3.1.1", 8333));
+
+    bench.run([&] {
+        const auto& address = addrman.Select();
+        assert(address.first.GetPort() > 0);
+    });
+}
+
+static void AddrManSelectByNetwork(benchmark::Bench& bench)
```

Comment:
It is excellent to see benchmark tests added! My only concern with this PR is that it will iterate through a lot of entries in addrman before finding one from the requested network. This in theory is `O(addrman size)`. And the benchmarks will help us assess that. So, it would be nice to compare, on the same, full addrman (e.g. 70k addresses):

* the speed of a regular `Select()`
* the speed of a `Select(network)` where the searched for address is near the end. In other words, worst case scenario, when ~70k addresses are iterated before finding the result.

Somehow I don't see that from the added benchmarks. I will play with them to see if that's possible (how to put an address "at the end"?).

Btw the results are unstable:
```
:wavy_dash: `AddrManSelectByNetwork` (Unstable with ~1.2 iters. Increase `minEpochIterations` to e.g. 12)
```

Reply:
This looks like a nice speedup in the case where we only have a few addresses - however, due to the constant overhead of building the shuffled list of buckets in the beginning I'd expect performance to go down a bit for the case where we have many addresses to choose from. Do you see this in your benchmark?
Why does this need `ALREADY_VISITED_AND_BORING`? If we do a for-loop through a pre-shuffled list of buckets instead of a while loop, doesn't that already guarantee that we visit each bucket at most once?

Reply:
@mzumsande,

> I'd expect performance to go down a bit for the case where we have many addresses to choose from. Do you see this in your benchmark?

Yes, that slows down a bit the fast `Select()` (any network is ok). I observed that and think it is ok because it is still super fast with or without the shuffling.

> Why does this need `ALREADY_VISITED_AND_BORING`?

The `for` loop could still repeat the whole array from the start if some address was found but was skipped. `ALREADY_VISITED_AND_BORING` is a further optimization to completely skip that bucket on the second and further passes through the array.

@amitiuttarwar, you are right that I modified the newly added `AddrManSelectByNetwork()` so that, in general, it tests the same thing as `AddrManSelect()` which already exists in `master`. I did that in order to be sure that everything else is _identical_: the way addrman is filled, `minEpochIterations(300)` and the `GetPort()` call. Way too many times I have been bitten by chasing noise in benchmarks, so I wanted to be sure that this is the only difference in the two things I am comparing:

```cpp
#if 0 // flip to 1 to test the other case
        addrman.Select(/*new_only=*/false);
#else
        addrman.Select(/*new_only=*/false, NET_I2P);
#endif
```

> also happy to discuss 2 further, but a question there - does it feel relevant to these patches to improve performance, or is it an orthogonal improvement

I think the performance improvement would be a nice addition to the work done by this PR (or maybe even a "must have"?)

> (since we see similar worst cases right now)?

Hmm? Now on `master` on mainnet it takes 0.18 microseconds and with this PR on mainnet it will take 6000 microseconds. That is the average case (averaged by the benchmark executing `Select` about 3500 times). The difference in the worst case is even more pronounced.

Reply:
> Hmm? Now on master on mainnet it takes 0.18 microseconds and with this PR on mainnet it will take 6000 microseconds. That is the average case (averaged by the benchmark executing Select about 3500 times). The difference in the worst case is even more pronounced.

We might be using slightly different terminology here: With "worst case", I mean the edge case where we have just one entry in AddrMan of the kind of address that we want. This is the benchmark `AddrManSelectFromAlmostEmpty` for unspecific selection and `AddrManSelectByNetwork` for network-specific selection. 
>#if 0 // flip to 1 to test the other case
        addrman.Select(/*new_only=*/false);
#else
        addrman.Select(/*new_only=*/false, NET_I2P);
#endif

This is not a meaningful comparison, because the first case is picking from an almost full AddrMan ("best case"), and the second case is the "worst case". If you replace `NET_I2P` by `NET_IPV4` in that example, I'd expect the network-specific performance to be basically identical to that of unspecific `Select()`, because AddrMan contains a single I2P address and ~50k IPV4 addresses in your setup.

Fluctuations due to RNG randomness within a given benchmark shouldn't be called "worst case" in my opinion, especially because there is no lower limit with the way `Select_()` currently works: for a given RNG seed, we might not visit the required bucket "forever" even if the probability for that will go to zero so quickly that it will never happen - there just is no distinct lower bound we could call "worst case" because it's probabilistic).

> Yes, that slows down a bit the fast `Select()` (any network is ok).

The crucial questions is by how much, because situations close to the "worst case" should be rather infrequent, while situations where we have multiple addresses to choose from should me much more common. With all the different methods of getting addresses, after a while of running your node you should at least know a couple dozens of addresses for each reachable network (even for CJDNS?!).
So there is the possibility that we might be optimizing for an edge case that is encountered very infrequently in the wild, while slightly slowing down the frequently encountered case at the same time. In this case, the change could be detrimental for performance overall!

I'm not saying this would be in this case here, but I want to do some more in-depth benchmarking myself later this week to convince myself a bit more that this would really be a clear performance improvement overall.

Reply:
I compared the performance of the PR branch with @vasild's [suggested optimization](https://github.com/bitcoin/bitcoin/pull/27214#discussion_r1132629513) ("VD") a bit more closely:

**Test 1**: Reasonably full addrman (15k addresses): `./bench_bitcoin -filter=AddrManSelect -min-time=5000`
=> The PR is much faster, since VD has a constant overhead due to the shuffling.
<details>
<summary>Details</summary>

PR:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              142.41 |        7,021,818.47 |    0.2% |      5.50 | `AddrManSelect`    
|              145.75 |        6,860,921.17 |    0.4% |      5.63 | `AddrManSelect`
|              145.91 |        6,853,404.83 |    1.4% |      5.55 | `AddrManSelect`
|              149.78 |        6,676,385.78 |    3.7% |      5.63 | `AddrManSelect`
|              144.37 |        6,926,771.77 |    0.3% |      5.49 | `AddrManSelect`

VD:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           12,716.43 |           78,638.44 |    0.1% |      5.49 | `AddrManSelect`
|           12,790.84 |           78,180.95 |    0.8% |      5.36 | `AddrManSelect`
|           12,884.37 |           77,613.41 |    0.7% |      5.53 | `AddrManSelect`
|           12,815.30 |           78,031.72 |    0.9% |      5.52 | `AddrManSelect`
|           12,961.91 |           77,149.10 |    1.3% |      5.55 | `AddrManSelect`
</details>


<br />

**Test 2**: A single address in AddrMan:  
`bench_bitcoin -filter=AddrManSelectFromAlmostEmpty -min-time=5000`
=> VD is faster by  a factor of 1.8
<details>
<summary>Details</summary>

PR:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           93,286.28 |           10,719.69 |    2.2% |      5.42 | `AddrManSelectFromAlmostEmpty`
|           90,781.42 |           11,015.47 |    1.0% |      5.52 | `AddrManSelectFromAlmostEmpty`
|           92,758.06 |           10,780.73 |    1.2% |      5.57 | `AddrManSelectFromAlmostEmpty`
|           90,026.20 |           11,107.88 |    0.7% |      5.52 | `AddrManSelectFromAlmostEmpty`
|           91,946.89 |           10,875.84 |    1.1% |      5.47 | `AddrManSelectFromAlmostEmpty`

VD:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           55,306.93 |           18,080.92 |    0.7% |      5.50 | `AddrManSelectFromAlmostEmpty`
|           55,209.86 |           18,112.71 |    0.4% |      5.48 | `AddrManSelectFromAlmostEmpty`
|           54,989.61 |           18,185.25 |    0.5% |      5.51 | `AddrManSelectFromAlmostEmpty`
|           56,727.70 |           17,628.07 |    2.1% |      5.57 | `AddrManSelectFromAlmostEmpty`
|           55,732.54 |           17,942.84 |    0.5% |      5.50 | `AddrManSelectFromAlmostEmpty`
</details>

**Test 3**: Three addresses in AddrMan:  set `NUM_SOURCES = 3`, `NUM_ADDRESSES_PER_SOURCE=1` then run 
`./bench_bitcoin -filter=AddrManSelect -min-time=5000`
=> PR is slightly faster.
<details>
<summary>Details</summary>

PR:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           30,034.99 |           33,294.50 |    0.7% |      5.51 | `AddrManSelect`
|           30,370.28 |           32,926.92 |    1.1% |      5.17 | `AddrManSelect`
|           30,574.52 |           32,706.98 |    2.1% |      5.20 | `AddrManSelect`
|           29,883.53 |           33,463.25 |    0.7% |      5.52 | `AddrManSelect`
|           29,944.31 |           33,395.33 |    0.5% |      5.52 | `AddrManSelect`

VD:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           34,126.18 |           29,303.02 |    0.7% |      5.30 | `AddrManSelect`
|           34,660.26 |           28,851.49 |    0.9% |      5.32 | `AddrManSelect`
|           34,075.81 |           29,346.33 |    0.7% |      5.30 | `AddrManSelect`
|           34,628.06 |           28,878.31 |    1.3% |      5.36 | `AddrManSelect`
|           34,300.57 |           29,154.03 |    0.6% |      5.31 | `AddrManSelect`
</details>

**Test 4**:
Query for a single network-specific address in an AddrMan filled with 15k IPV4 addresses we don't want
`./bench_bitcoin -filter=AddrManSelectByNetwork -min-time=500`
=> VD is faster by a factor of ~2.1.
<details>
<summary>Details</summary>

PR:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|        1,256,860.84 |              795.63 |    4.4% |      5.40 | `AddrManSelectByNetwork`
|        1,269,322.70 |              787.82 |    2.6% |      5.50 | `AddrManSelectByNetwork`
|        1,242,639.25 |              804.74 |    1.9% |      5.56 | `AddrManSelectByNetwork`
|        1,246,311.65 |              802.37 |    2.9% |      5.69 | `AddrManSelectByNetwork`
|        1,265,535.68 |              790.18 |    3.4% |      5.59 | `AddrManSelectByNetwork`

VD
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          560,818.30 |            1,783.11 |    1.0% |      5.56 | `AddrManSelectByNetwork`
|          558,351.91 |            1,790.99 |    0.9% |      5.53 | `AddrManSelectByNetwork`
|          600,536.33 |            1,665.18 |    4.8% |      5.38 | `AddrManSelectByNetwork`
|          569,208.08 |            1,756.83 |    1.7% |      5.45 | `AddrManSelectByNetwork`
|          581,860.03 |            1,718.63 |    2.3% |      5.47 | `AddrManSelectByNetwork`
</details>

**Test 5**:
Query for 20 network-specific address in an AddrMan filled with 15k IPV4 addresses we don't want (changing the code a bit)
`./bench_bitcoin -filter=AddrManSelectByNetwork -min-time=500`
=> Performance is approximately the same.
<details>
<summary>Details</summary>

PR:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          109,388.10 |            9,141.76 |    1.6% |      5.40 | `AddrManSelectByNetwork`
|          108,804.63 |            9,190.79 |    2.1% |      5.56 | `AddrManSelectByNetwork`
|          111,333.16 |            8,982.05 |    2.0% |      5.61 | `AddrManSelectByNetwork`
|          110,827.61 |            9,023.02 |    1.4% |      5.55 | `AddrManSelectByNetwork`
|          122,518.06 |            8,162.06 |    1.4% |      5.47 | `AddrManSelectByNetwork`

VD
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          106,806.78 |            9,362.70 |    2.5% |      5.51 | `AddrManSelectByNetwork`
|          109,221.66 |            9,155.69 |    3.1% |      5.59 | `AddrManSelectByNetwork`
|          107,870.77 |            9,270.35 |    3.8% |      5.41 | `AddrManSelectByNetwork`
|          105,131.17 |            9,511.93 |    0.5% |      5.47 | `AddrManSelectByNetwork`
|          106,080.47 |            9,426.81 |    1.2% |      5.48 | `AddrManSelectByNetwork`

</details>

In conclusion, my results indicate that the suggested optimization is only faster in some specific scenarios because the pre-shuffling of the buckets adds a constant overhead that is not neglible:
In the case of non-network specific queries (which should be the bulk even after #17213 ), the PR is already faster when `AddrMan` has more than 3 addresses (which should virtually always be the case!)
In the case of network-specific queries, the tipping point is (because of the additional cost of looking up addrs in buckets that we can't use) at ~20 network-specific addresses (with 15k IPV4 addresses). This is a bit higher, but I would imagine that we should get more than 20 addresses of most alternative networks rather quickly.
To summarize, my results indicate that adding pre-shuffling AddrMan buckets to prevent repeated visits appears to be no clear performance improvement - in most cases, performance would go down.

@vasild: Do you see similar numbers?

Reply:
About the optimization in https://github.com/bitcoin/bitcoin/pull/27214#discussion_r1132629513, couldn't we simply use a vector to store the visited buckets?

just an example:
```cpp
  std::vector<int> buckets_visited;
  while (1) {
      // Pick a bucket, and an initial position in that bucket.
      int bucket = insecure_rand.randrange(bucket_count);
      if (std::find(buckets_visited.begin(), buckets_visited.end(), bucket) != buckets_visited.end()) {
          continue;
      } else {
          buckets_visited.push_back(bucket);
      }
```

Also, I ran same tests that @mzumsande did and got similar results.

Reply:
@mzumsande, I think you are missing my point, let me try to rephrase it: now on `master`, from `ThreadOpenConnections()` we call `Select()`. With the parent of this PR, we will be calling `Select(network)` which will bring some security (good) + some slowdown (bad).

Try to asses how much is the slowdown. Use a full addrman (15k is not full) because that is what people out there are running on. If the slowdown is ok, then no further improvements are necessary on this PR or its parent.

What is your "Test 5" doing? Does it call 20 times `Select(network)`?

Further thoughts if an improvement of performance of `Select(network)` is deemed necessary:
* The shuffling can be done only if `network` is provided, to avoid slowing down of `Select()` without a network.
* The shuffling can be removed altogether and just marking the buckets as boring to be left. This should still improve the 600k iterations in the worst case I observed.
* The ultimate beast solution would be to have an index per network, e.g. `std::unordered_set<size_t> positions_in_vRandom_of_i2p_addresses`.

I will try to compare on a snapshot of an addrman from my public full node, not on artificially filled addrman from the bench. Last time I checked it had ~70k addresses and also it will have at least a bunch of addresses from each network, not just 1 address. That would be more representative.

Reply:
@brunoerg, yes, that should still bring some improvement. But only push it into the vector if we have visited all addresses in that bucket and found 0 from the requested network because it may happen that we visit a bucket with an address from the requested network, but only pick it up on some of the subsequent iterations, not from the first one. Change it to `std::unordered_set` so that lookup is fast (`O(1)`).

Reply:
> Last time I checked it had ~70k addresses and also it will have at least a bunch of addresses from each network, not just 1 address.

"bunch of addresses from each network", how much? Last time I checked my node (running with multiple networks) I remembered that the number of i2p addresses were so small compared to clearnet, for example.

Reply:
> What is your "Test 5" doing? Does it call 20 times `Select(network)`?

No, it calls `Select(network)` for an addrman that has 20 addresses of the selected network.

I'll try to explain the idea behind my benchmarks better: 
Starting point is the observation that the suggested change makes `Select` slower if we already know a lot of addresses of the desired type, while making it faster if addrman has very few addresses of this type. This is true for botch `Select()` and `Select(network)`, which is why it is not obvious at first sight if the suggestion is an overall performance improvement or a degradation. 

So my idea was to find the cutoff - how many addresses do we need such that both algorithms are equally fast - in situations where we have more addrs than this, the change would be a slowdown, when we have less it's a speedup. This cutoff turned out to be very small for `Select()` (3 addresses), while being a bit larger for `Select(network)` - it was approximately 20. It would be helpful to know how things change with a full addrman!

So if we used an addrman with counts similar to those reported by @jonatack, my results would suggest that the pre-shufffling change suggested [here](https://github.com/bitcoin/bitcoin/pull/27214#discussion_r1132629513) would be a speedup for CJDNS-specific queries, while being a slowdown for clearnet, onion, i2p and unspecific queries (all compared to this PR).

Reply:
I ran this on a mainnet `peers.dat`:
```
addrman size total: 72947
addrman size IPv4:  47252
addrman size IPv6:  11491
addrman size Tor:   13305
addrman size I2P:   893
addrman size CJDNS: 6
```

<details>
<summary>[patch] How to load /tmp/peers.dat from bench_bitcoin</summary>

```diff
diff --git i/src/bench/addrman.cpp w/src/bench/addrman.cpp
index 8a5cab443f..b8ec4a4071 100644
--- i/src/bench/addrman.cpp
+++ w/src/bench/addrman.cpp
@@ -1,17 +1,22 @@
 // Copyright (c) 2020-2022 The Bitcoin Core developers
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <addrdb.h>
 #include <addrman.h>
 #include <bench/bench.h>
+#include <chainparams.h>
+#include <chainparamsbase.h>
 #include <netbase.h>
 #include <netgroup.h>
 #include <random.h>
 #include <util/check.h>
+#include <util/system.h>
 #include <util/time.h>
+#include <util/translation.h>
 
 #include <optional>
 #include <vector>
 
 /* A "source" is a source address from which we have received a bunch of other addresses. */
 
@@ -21,12 +26,23 @@ static constexpr size_t NUM_ADDRESSES_PER_SOURCE = 256;
 static NetGroupManager EMPTY_NETGROUPMAN{std::vector<bool>()};
 static constexpr uint32_t ADDRMAN_CONSISTENCY_CHECK_RATIO{0};
 
 static std::vector<CAddress> g_sources;
 static std::vector<std::vector<CAddress>> g_addresses;
 
+std::unique_ptr<AddrMan> LoadAddrmanFromPeersDat()
+{
+    SelectBaseParams("main");
+    SelectParams("main");
+    gArgs.ForceSetArg("-datadir", "/tmp");
+    std::unique_ptr<AddrMan> addrman;
+    auto err = LoadAddrman(EMPTY_NETGROUPMAN, gArgs, addrman);
+    assert(!err.has_value());
+    return addrman;
+}
+
 static void CreateAddresses()
 {
     if (g_sources.size() > 0) { // already created
         return;
     }
 
@@ -69,19 +85,12 @@ static void FillAddrMan(AddrMan& addrman)
 {
     CreateAddresses();
 
     AddAddressesToAddrMan(addrman);
 }
 
-static CNetAddr ResolveIP(const std::string& ip)
-{
-    CNetAddr addr;
-    LookupHost(ip, addr, false);
-    return addr;
-}
-
 static CService ResolveService(const std::string& ip, uint16_t port = 0)
 {
     CService serv;
     Lookup(ip, serv, port, false);
     return serv;
 }
@@ -125,26 +134,24 @@ static void AddrManSelectFromAlmostEmpty(benchmark::Bench& bench)
         (void)addrman.Select();
     });
 }
 
 static void AddrManSelectByNetwork(benchmark::Bench& bench)
 {
-    AddrMan addrman{EMPTY_NETGROUPMAN, /*deterministic=*/false, ADDRMAN_CONSISTENCY_CHECK_RATIO};
-
-    // add single I2P address to new table
-    CService i2p_service;
-    i2p_service.SetSpecial("udhdrtrcetjm5sxzskjyr5ztpeszydbh4dpl3pl4utgqqw2v4jna.b32.i2p");
-    CAddress i2p_address(i2p_service, NODE_NONE);
-    i2p_address.nTime = Now<NodeSeconds>();
-    CNetAddr source = ResolveIP("252.2.2.2");
-    addrman.Add({i2p_address}, source);
-
-    FillAddrMan(addrman);
+    auto addrman = LoadAddrmanFromPeersDat();
+    std::cout << "addrman size total: " << addrman->Size() << std::endl
+              << "addrman size IPv4:  " << addrman->Size(NET_IPV4) << std::endl
+              << "addrman size IPv6:  " << addrman->Size(NET_IPV6) << std::endl
+              << "addrman size Tor:   " << addrman->Size(NET_ONION) << std::endl
+              << "addrman size I2P:   " << addrman->Size(NET_I2P) << std::endl
+              << "addrman size CJDNS: " << addrman->Size(NET_CJDNS) << std::endl;
 
     bench.run([&] {
-        (void)addrman.Select(/*new_only=*/false, NET_I2P);
+        //addrman->Select(/*new_only=*/false);
+        //addrman->Select(/*new_only=*/false, NET_I2P);
+        addrman->Select(/*new_only=*/false, NET_CJDNS);
     });
 }
 
 static void AddrManGetAddr(benchmark::Bench& bench)
 {
     AddrMan addrman{EMPTY_NETGROUPMAN, /*deterministic=*/false, ADDRMAN_CONSISTENCY_CHECK_RATIO};
```
</details>

Results:

* `Select(/*new_only=*/false)`: 0.265 microseconds, this is what `master` is doing before this PR

After this PR:

* `Select(/*new_only=*/false, NET_IPV4)`: 0.5 microseconds
* `Select(/*new_only=*/false, NET_IPV6)`:  1.7 microseconds
* `Select(/*new_only=*/false, NET_ONION)`: 3.1 microseconds
* `Select(/*new_only=*/false, NET_I2P)`: 27 microseconds
* `Select(/*new_only=*/false, NET_CJDNS)`: 1600 microseconds

Is this slowdown acceptable? Maybe yes. We don't call `Select` in a tight loop, so maybe even the 1600 microseconds is ok.

What happens if we mark and subsequently skip buckets we have seen to contain 0 interesting addresses, without the shuffling? I guess that should bring most of the improvement from the above patch without the downside of the shuffling. Here is the change on top of this PR:

<details>
<summary>[patch] Skip boring</summary>

```diff
diff --git i/src/addrman.cpp w/src/addrman.cpp
index cdfd079fcd..46f4de99be 100644
--- i/src/addrman.cpp
+++ w/src/addrman.cpp
@@ -747,15 +747,19 @@ std::pair<CAddress, NodeSeconds> AddrManImpl::Select_(bool new_only, std::option
     }
 
     const int bucket_count{search_tried ? ADDRMAN_TRIED_BUCKET_COUNT : ADDRMAN_NEW_BUCKET_COUNT};
 
     // Loop through the addrman table until we find an appropriate entry
     double chance_factor = 1.0;
+    std::unordered_set<int> already_visited_and_boring_buckets;
     while (1) {
         // Pick a bucket, and an initial position in that bucket.
         int bucket = insecure_rand.randrange(bucket_count);
+        if (already_visited_and_boring_buckets.count(bucket) > 0) {
+            continue;
+        }
         int initial_position = insecure_rand.randrange(ADDRMAN_BUCKET_SIZE);
 
         // Iterate over the positions of that bucket, starting at the initial one,
         // and looping around.
         int i;
         for (i = 0; i < ADDRMAN_BUCKET_SIZE; ++i) {
@@ -770,14 +774,19 @@ std::pair<CAddress, NodeSeconds> AddrManImpl::Select_(bool new_only, std::option
                 } else {
                     break;
                 }
             }
         }
 
-        // If the bucket is entirely empty, start over with a (likely) different one.
-        if (i == ADDRMAN_BUCKET_SIZE) continue;
+        // Start over with a different bucket if this one is entirely empty or
+        // specific network was requested and it does not contain any addresses
+        // from that network.
+        if (i == ADDRMAN_BUCKET_SIZE) {
+            already_visited_and_boring_buckets.insert(bucket);
+            continue;
+        }
 
         // Find the entry to return.
         int position = (initial_position + i) % ADDRMAN_BUCKET_SIZE;
         int nId = GetEntry(search_tried, bucket, position);
         const auto it_found{mapInfo.find(nId)};
         assert(it_found != mapInfo.end());
```
</details>

Result for `NET_CJDNS` where the slowdown is most pronounced: 1200 microseconds (vs 1600). Those numbers are averaged over many runs and they do not tell the whole story. Here are the distributions without "skip boring" (the height of a bar indicates how many addresses were checked before finding the result, e.g. if the bar that spans between 2500 and 3500 is tall 280, it means 280 of the runs checked 2500-3500 addresses before finding the result).

![1](https://user-images.githubusercontent.com/266751/228896184-9e5fe576-8c0f-44d2-9f3d-f7061ac936e3.png)

with "skip boring":

![2](https://user-images.githubusercontent.com/266751/228896576-20b8cc6d-a121-4836-b727-adc04079a55d.png)

Reply:
I don't see why 1.6 milliseconds to choose an address would be a concern.

Reply:
A simpler approach to skipping already tried buckets might be something like this:

```c++
    int n = 1 << 8; // power of 2
    int initial_pos = random(n);
    int step = random(n/2) * 2 + 1; // odd number
    int i = initial_pos;
    do {
        if (check_bucket(i)) break;
        i = (i + step) % n; // oops, try a new bucket
    } while (i != initial_pos);
```

That doesn't need a full shuffle, still hits every bucket, and only hits each bucket once (provided step and n are co-prime).

Based on vasild's stats above, I think this is fine to merge without any improvement here, but it might be interesting to rebench with this approach -- performance should benefit a little from needing less randomness as well as avoiding repeated work...

If we wanted to making things actually reliably fast for cjdns-like networks lost in a sea of ipv4 addresses, then having separate addrman tables for each network might be worth exploring. That would probably help prevent an attacker from blocking valid ipv4 addresses with a mass of valid-but-sybiled tor addresses, which might be worthwhile.